cmd/disk-cache-backend.go

```diff
@@ -0,0 +1,565 @@
+	for _, hashStr := range info.Hashes {
+		h, err := hex.DecodeString(hashStr)
+		if err != nil {
+			return err
+		}
+		hashes = append(hashes, h)
+	}
+	c.Hashes = hashes
+	return nil
+}
+
+// Represents the cache metadata struct
+type cacheMeta struct {
+	Version string `json:"version"`
+	// checksums of blocks on disk.
+	Checksum FSChecksumInfoV1 `json:"checksum,omitempty"`
+	// Metadata map for current object.
+	Meta map[string]string `json:"meta,omitempty"`
+	// parts info for current object
+	Parts []ObjectPartInfo `json:"parts,omitempty"`
+}
+
+func (m *cacheMeta) ToObjectInfo(bucket, object string, fi os.FileInfo) (o ObjectInfo) {
+	if len(m.Meta) == 0 {
+		m.Meta = make(map[string]string)
+	}
+
+	o = ObjectInfo{
+		Bucket: bucket,
+		Name:   object,
+	}
+
+	// We set file info only if its valid.
+	o.ModTime = timeSentinel
+	if fi != nil {
+		o.ModTime = fi.ModTime()
+		o.Size = fi.Size()
+		if fi.IsDir() {
+			// Directory is always 0 bytes in S3 API, treat it as such.
+			o.Size = 0
+			o.IsDir = fi.IsDir()
+		}
+	}
+
+	o.ETag = extractETag(m.Meta)
+	o.ContentType = m.Meta["content-type"]
+	o.ContentEncoding = m.Meta["content-encoding"]
+	if storageClass, ok := m.Meta[amzStorageClass]; ok {
+		o.StorageClass = storageClass
+	} else {
+		o.StorageClass = globalMinioDefaultStorageClass
+	}
+	var (
+		t time.Time
+		e error
+	)
+	if exp, ok := m.Meta["expires"]; ok {
+		if t, e = time.Parse(http.TimeFormat, exp); e == nil {
+			o.Expires = t.UTC()
+		}
+	}
+	// etag/md5Sum has already been extracted. We need to
+	// remove to avoid it from appearing as part of
+	o.UserDefined = cleanMetadata(m.Meta)
+	return o
+}
+
+// disk cache
+type diskCache struct {
+	dir             string // caching directory
+	maxDiskUsagePct int    // max usage in %
+	expiry          int    // cache expiry in days
+	// to manage cache operations
+	nsMutex *nsLockMap
+	// mark false if drive is offline
+	online bool
+	// mutex to protect updates to online variable
+	onlineMutex *sync.RWMutex
+	// purge() listens on this channel to start the cache-purge process
+	purgeChan chan struct{}
+}
+
+// Inits the disk cache dir if it is not init'ed already.
+func newdiskCache(dir string, expiry int, maxDiskUsagePct int) (*diskCache, error) {
+	if err := os.MkdirAll(dir, 0777); err != nil {
+		return nil, fmt.Errorf("Unable to initialize '%s' dir, %s", dir, err)
+	}
+
+	if expiry == 0 {
+		expiry = globalCacheExpiry
+	}
+
+	cache := diskCache{
+		dir:             dir,
+		expiry:          expiry,
+		maxDiskUsagePct: maxDiskUsagePct,
+		purgeChan:       make(chan struct{}),
+		online:          true,
+		onlineMutex:     &sync.RWMutex{},
+		nsMutex:         newNSLock(false),
+	}
+	return &cache, nil
+}
+
+// migrate cache contents from old cacheFS format to new format
+func (c *diskCache) migrateOldCache() error {
+	ctx := context.Background()
+	oldCacheBucketsPath := path.Join(c.dir, minioMetaBucket, "buckets")
+	if _, err := os.Stat(oldCacheBucketsPath); err != nil {
```

Comment:
Note that migration should be based on format.json's version number (i.e that is the purpose of version in format.json)

Reply:
Once migrated, the version number should be made 2

Reply:
done